In [51]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [52]:
data = pd.read_csv('sales_data.csv') # read csv file 
data # displays first 5 rows and last 5 rows

Date  Day     Month  Year  Customer_Age       Age_Group  \
0       2013-11-26   26  November  2013            19     Youth (<25)   
1       2015-11-26   26  November  2015            19     Youth (<25)   
2       2014-03-23   23     March  2014            49  Adults (35-64)   
3       2016-03-23   23     March  2016            49  Adults (35-64)   
4       2014-05-15   15       May  2014            47  Adults (35-64)   
...            ...  ...       ...   ...           ...             ...   
113031  2016-04-12   12     April  2016            41  Adults (35-64)   
113032  2014-04-02    2     April  2014            18     Youth (<25)   
113033  2016-04-02    2     April  2016            18     Youth (<25)   
113034  2014-03-04    4     March  2014            37  Adults (35-64)   
113035  2016-03-04    4     March  2016            37  Adults (35-64)   

       Customer_Gender         Country             State Product_Category  \
0                    M          Canada  British Columbia      Accessories   
1                    M          Canada  British Columbia      Accessories   
2                    M       Australia   New South Wales      Accessories   
3                    M       Australia   New South Wales      Accessories   
4                    F       Australia   New South Wales      Accessories   
...                ...             ...               ...              ...   
113031               M  United Kingdom           England         Clothing   
113032               M       Australia        Queensland         Clothing   
113033               M       Australia        Queensland         Clothing   
113034               F          France     Seine (Paris)         Clothing   
113035               F          France     Seine (Paris)         Clothing   

       Sub_Category              Product  Order_Quantity  Unit_Cost  \
0        Bike Racks  Hitch Rack - 4-Bike               8         45   
1        Bike Racks  Hitch Rack - 4-Bike               8         45   
2        Bike Racks  Hitch Rack - 4-Bike              23         45   
3        Bike Racks  Hitch Rack - 4-Bike              20         45   
4        Bike Racks  Hitch Rack - 4-Bike               4         45   
...             ...                  ...             ...        ...   
113031        Vests      Classic Vest, S               3         24   
113032        Vests      Classic Vest, M              22         24   
113033        Vests      Classic Vest, M              22         24   
113034        Vests      Classic Vest, L              24         24   
113035        Vests      Classic Vest, L              23         24   

        Unit_Price  Profit  Cost  Revenue  
0              120     590   360      950  
1              120     590   360      950  
2              120    1366  1035     2401  
3              120    1188   900     2088  
4              120     238   180      418  
...            ...     ...   ...      ...  
113031          64     112    72      184  
113032          64     655   528     1183  
113033          64     655   528     1183  
113034          64     684   576     1260  
113035          64     655   552     1207  

[113036 rows x 18 columns]

## Handling Missing Values

In [53]:
data.isnull().sum()

Date                0
Day                 0
Month               0
Year                0
Customer_Age        0
Age_Group           0
Customer_Gender     0
Country             0
State               0
Product_Category    0
Sub_Category        0
Product             0
Order_Quantity      0
Unit_Cost           0
Unit_Price          0
Profit              0
Cost                0
Revenue             0
dtype: int64

-  hence no null values in Sales Data

## Removing Duplicate Rows

In [54]:
data.duplicated().any() # returns duplicate value is present or not in the form of true or false

True

In [55]:
data.duplicated().sum() # sums up the total number of duplicate values

1000

In [56]:
data = data.drop_duplicates() # removes the duplicted rows

In [57]:
data # to display the data

Date  Day     Month  Year  Customer_Age       Age_Group  \
0       2013-11-26   26  November  2013            19     Youth (<25)   
1       2015-11-26   26  November  2015            19     Youth (<25)   
2       2014-03-23   23     March  2014            49  Adults (35-64)   
3       2016-03-23   23     March  2016            49  Adults (35-64)   
4       2014-05-15   15       May  2014            47  Adults (35-64)   
...            ...  ...       ...   ...           ...             ...   
113031  2016-04-12   12     April  2016            41  Adults (35-64)   
113032  2014-04-02    2     April  2014            18     Youth (<25)   
113033  2016-04-02    2     April  2016            18     Youth (<25)   
113034  2014-03-04    4     March  2014            37  Adults (35-64)   
113035  2016-03-04    4     March  2016            37  Adults (35-64)   

       Customer_Gender         Country             State Product_Category  \
0                    M          Canada  British Columbia      Accessories   
1                    M          Canada  British Columbia      Accessories   
2                    M       Australia   New South Wales      Accessories   
3                    M       Australia   New South Wales      Accessories   
4                    F       Australia   New South Wales      Accessories   
...                ...             ...               ...              ...   
113031               M  United Kingdom           England         Clothing   
113032               M       Australia        Queensland         Clothing   
113033               M       Australia        Queensland         Clothing   
113034               F          France     Seine (Paris)         Clothing   
113035               F          France     Seine (Paris)         Clothing   

       Sub_Category              Product  Order_Quantity  Unit_Cost  \
0        Bike Racks  Hitch Rack - 4-Bike               8         45   
1        Bike Racks  Hitch Rack - 4-Bike               8         45   
2        Bike Racks  Hitch Rack - 4-Bike              23         45   
3        Bike Racks  Hitch Rack - 4-Bike              20         45   
4        Bike Racks  Hitch Rack - 4-Bike               4         45   
...             ...                  ...             ...        ...   
113031        Vests      Classic Vest, S               3         24   
113032        Vests      Classic Vest, M              22         24   
113033        Vests      Classic Vest, M              22         24   
113034        Vests      Classic Vest, L              24         24   
113035        Vests      Classic Vest, L              23         24   

        Unit_Price  Profit  Cost  Revenue  
0              120     590   360      950  
1              120     590   360      950  
2              120    1366  1035     2401  
3              120    1188   900     2088  
4              120     238   180      418  
...            ...     ...   ...      ...  
113031          64     112    72      184  
113032          64     655   528     1183  
113033          64     655   528     1183  
113034          64     684   576     1260  
113035          64     655   552     1207  

[112036 rows x 18 columns]

- Hence the duplicated rows is droped before dropping the duplicated rows the total number of rows was 113036 after dropping the duplicated rows the total number of rows becomes 112036

##  Standardize text values 

In [58]:
# Clean 'Customer_Gender' and 'Country' columns
data['Customer_Gender'] = data['Customer_Gender'].str.strip().str.lower()
data['Country'] = data['Country'].str.strip().str.title()

- **.str.strip()** --> removes leading/trailing spaces
- **.str.lower()** → makes all text lowercase (e.g., "M", "F" → "m", "f")
- **.str.upper()** → makes all text uppercase
- **.str.title()** → capitalizes each word (e.g., "united states" → "United States")

In [62]:
data

Date  Day     Month  Year  Customer_Age       Age_Group  \
0       2013-11-26   26  November  2013            19     Youth (<25)   
1       2015-11-26   26  November  2015            19     Youth (<25)   
2       2014-03-23   23     March  2014            49  Adults (35-64)   
3       2016-03-23   23     March  2016            49  Adults (35-64)   
4       2014-05-15   15       May  2014            47  Adults (35-64)   
...            ...  ...       ...   ...           ...             ...   
113031  2016-04-12   12     April  2016            41  Adults (35-64)   
113032  2014-04-02    2     April  2014            18     Youth (<25)   
113033  2016-04-02    2     April  2016            18     Youth (<25)   
113034  2014-03-04    4     March  2014            37  Adults (35-64)   
113035  2016-03-04    4     March  2016            37  Adults (35-64)   

       Customer_Gender         Country             State Product_Category  \
0                    m          Canada  British Columbia      Accessories   
1                    m          Canada  British Columbia      Accessories   
2                    m       Australia   New South Wales      Accessories   
3                    m       Australia   New South Wales      Accessories   
4                    f       Australia   New South Wales      Accessories   
...                ...             ...               ...              ...   
113031               m  United Kingdom           England         Clothing   
113032               m       Australia        Queensland         Clothing   
113033               m       Australia        Queensland         Clothing   
113034               f          France     Seine (Paris)         Clothing   
113035               f          France     Seine (Paris)         Clothing   

       Sub_Category              Product  Order_Quantity  Unit_Cost  \
0        Bike Racks  Hitch Rack - 4-Bike               8         45   
1        Bike Racks  Hitch Rack - 4-Bike               8         45   
2        Bike Racks  Hitch Rack - 4-Bike              23         45   
3        Bike Racks  Hitch Rack - 4-Bike              20         45   
4        Bike Racks  Hitch Rack - 4-Bike               4         45   
...             ...                  ...             ...        ...   
113031        Vests      Classic Vest, S               3         24   
113032        Vests      Classic Vest, M              22         24   
113033        Vests      Classic Vest, M              22         24   
113034        Vests      Classic Vest, L              24         24   
113035        Vests      Classic Vest, L              23         24   

        Unit_Price  Profit  Cost  Revenue  
0              120     590   360      950  
1              120     590   360      950  
2              120    1366  1035     2401  
3              120    1188   900     2088  
4              120     238   180      418  
...            ...     ...   ...      ...  
113031          64     112    72      184  
113032          64     655   528     1183  
113033          64     655   528     1183  
113034          64     684   576     1260  
113035          64     655   552     1207  

[112036 rows x 18 columns]

## Convert date formats 

In [63]:
# Step 1: Ensure the column is in datetime format
data['Date'] = pd.to_datetime(data['Date'], errors='coerce')

# Step 2: Convert to dd-mm-yyyy format (as string)
data['Date'] = data['Date'].dt.strftime('%d-%m-%Y')

In [64]:
data

Date  Day     Month  Year  Customer_Age       Age_Group  \
0       26-11-2013   26  November  2013            19     Youth (<25)   
1       26-11-2015   26  November  2015            19     Youth (<25)   
2       23-03-2014   23     March  2014            49  Adults (35-64)   
3       23-03-2016   23     March  2016            49  Adults (35-64)   
4       15-05-2014   15       May  2014            47  Adults (35-64)   
...            ...  ...       ...   ...           ...             ...   
113031  12-04-2016   12     April  2016            41  Adults (35-64)   
113032  02-04-2014    2     April  2014            18     Youth (<25)   
113033  02-04-2016    2     April  2016            18     Youth (<25)   
113034  04-03-2014    4     March  2014            37  Adults (35-64)   
113035  04-03-2016    4     March  2016            37  Adults (35-64)   

       Customer_Gender         Country             State Product_Category  \
0                    m          Canada  British Columbia      Accessories   
1                    m          Canada  British Columbia      Accessories   
2                    m       Australia   New South Wales      Accessories   
3                    m       Australia   New South Wales      Accessories   
4                    f       Australia   New South Wales      Accessories   
...                ...             ...               ...              ...   
113031               m  United Kingdom           England         Clothing   
113032               m       Australia        Queensland         Clothing   
113033               m       Australia        Queensland         Clothing   
113034               f          France     Seine (Paris)         Clothing   
113035               f          France     Seine (Paris)         Clothing   

       Sub_Category              Product  Order_Quantity  Unit_Cost  \
0        Bike Racks  Hitch Rack - 4-Bike               8         45   
1        Bike Racks  Hitch Rack - 4-Bike               8         45   
2        Bike Racks  Hitch Rack - 4-Bike              23         45   
3        Bike Racks  Hitch Rack - 4-Bike              20         45   
4        Bike Racks  Hitch Rack - 4-Bike               4         45   
...             ...                  ...             ...        ...   
113031        Vests      Classic Vest, S               3         24   
113032        Vests      Classic Vest, M              22         24   
113033        Vests      Classic Vest, M              22         24   
113034        Vests      Classic Vest, L              24         24   
113035        Vests      Classic Vest, L              23         24   

        Unit_Price  Profit  Cost  Revenue  
0              120     590   360      950  
1              120     590   360      950  
2              120    1366  1035     2401  
3              120    1188   900     2088  
4              120     238   180      418  
...            ...     ...   ...      ...  
113031          64     112    72      184  
113032          64     655   528     1183  
113033          64     655   528     1183  
113034          64     684   576     1260  
113035          64     655   552     1207  

[112036 rows x 18 columns]

## Renaming column headers

In [65]:
data.columns = data.columns.str.strip().str.lower().str.replace(' ', '_')

In [66]:
data 

date  day     month  year  customer_age       age_group  \
0       26-11-2013   26  November  2013            19     Youth (<25)   
1       26-11-2015   26  November  2015            19     Youth (<25)   
2       23-03-2014   23     March  2014            49  Adults (35-64)   
3       23-03-2016   23     March  2016            49  Adults (35-64)   
4       15-05-2014   15       May  2014            47  Adults (35-64)   
...            ...  ...       ...   ...           ...             ...   
113031  12-04-2016   12     April  2016            41  Adults (35-64)   
113032  02-04-2014    2     April  2014            18     Youth (<25)   
113033  02-04-2016    2     April  2016            18     Youth (<25)   
113034  04-03-2014    4     March  2014            37  Adults (35-64)   
113035  04-03-2016    4     March  2016            37  Adults (35-64)   

       customer_gender         country             state product_category  \
0                    m          Canada  British Columbia      Accessories   
1                    m          Canada  British Columbia      Accessories   
2                    m       Australia   New South Wales      Accessories   
3                    m       Australia   New South Wales      Accessories   
4                    f       Australia   New South Wales      Accessories   
...                ...             ...               ...              ...   
113031               m  United Kingdom           England         Clothing   
113032               m       Australia        Queensland         Clothing   
113033               m       Australia        Queensland         Clothing   
113034               f          France     Seine (Paris)         Clothing   
113035               f          France     Seine (Paris)         Clothing   

       sub_category              product  order_quantity  unit_cost  \
0        Bike Racks  Hitch Rack - 4-Bike               8         45   
1        Bike Racks  Hitch Rack - 4-Bike               8         45   
2        Bike Racks  Hitch Rack - 4-Bike              23         45   
3        Bike Racks  Hitch Rack - 4-Bike              20         45   
4        Bike Racks  Hitch Rack - 4-Bike               4         45   
...             ...                  ...             ...        ...   
113031        Vests      Classic Vest, S               3         24   
113032        Vests      Classic Vest, M              22         24   
113033        Vests      Classic Vest, M              22         24   
113034        Vests      Classic Vest, L              24         24   
113035        Vests      Classic Vest, L              23         24   

        unit_price  profit  cost  revenue  
0              120     590   360      950  
1              120     590   360      950  
2              120    1366  1035     2401  
3              120    1188   900     2088  
4              120     238   180      418  
...            ...     ...   ...      ...  
113031          64     112    72      184  
113032          64     655   528     1183  
113033          64     655   528     1183  
113034          64     684   576     1260  
113035          64     655   552     1207  

[112036 rows x 18 columns]

- **.str.strip()** → removes leading/trailing whitespace
- **.str.lower()** → converts to lowercase
- **.str.replace(' ', '_')** → replaces spaces with underscores

## Check and fix datatypes

In [67]:
data.dtypes # check for current data types

date                object
day                  int64
month               object
year                 int64
customer_age         int64
age_group           object
customer_gender     object
country             object
state               object
product_category    object
sub_category        object
product             object
order_quantity       int64
unit_cost            int64
unit_price           int64
profit               int64
cost                 int64
revenue              int64
dtype: object

In [68]:
data['customer_age'] = data['customer_age'].astype(int) # converts the specified column value to integer

In [69]:
data['date'] = pd.to_datetime(data['date'], errors='coerce') # converts the specified column to datetime format

In [70]:
data['product'] = data['product'].astype(str) # converts the specified column to string

In [71]:
data.dtypes # recheck the datatypes after conversion

date                datetime64[ns]
day                          int64
month                       object
year                         int64
customer_age                 int32
age_group                   object
customer_gender             object
country                     object
state                       object
product_category            object
sub_category                object
product                     object
order_quantity               int64
unit_cost                    int64
unit_price                   int64
profit                       int64
cost                         int64
revenue                      int64
dtype: object